x-pack/plugins/security/server/authentication/authenticator.ts

```diff
@@ -0,0 +1,406 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import {
+  SessionStorageFactory,
+  SessionStorage,
+  KibanaRequest,
+  LoggerFactory,
+  Logger,
+  HttpServiceSetup,
+  ClusterClient,
+} from '../../../../../src/core/server';
+import { ConfigType } from '../config';
+import { getErrorStatusCode } from '../errors';
+
+import {
+  AuthenticationProviderOptions,
+  AuthenticationProviderSpecificOptions,
+  BaseAuthenticationProvider,
+  BasicAuthenticationProvider,
+  KerberosAuthenticationProvider,
+  SAMLAuthenticationProvider,
+  TokenAuthenticationProvider,
+  OIDCAuthenticationProvider,
+  isSAMLRequestQuery,
+} from './providers';
+import { AuthenticationResult } from './authentication_result';
+import { DeauthenticationResult } from './deauthentication_result';
+import { Tokens } from './tokens';
+
+/**
+ * The shape of the session that is actually stored in the cookie.
+ */
+export interface ProviderSession {
+  /**
+   * Name/type of the provider this session belongs to.
+   */
+  provider: string;
+
+  /**
+   * The Unix time in ms when the session should be considered expired. If `null`, session will stay
+   * active until the browser is closed.
+   */
+  expires: number | null;
+
+  /**
+   * Session value that is fed to the authentication provider. The shape is unknown upfront and
+   * entirely determined by the authentication provider that owns the current session.
+   */
+  state: unknown;
+}
+
+/**
+ * The shape of the login attempt.
+ */
+export interface ProviderLoginAttempt {
+  /**
+   * Name/type of the provider this login attempt is targeted for.
+   */
+  provider: string;
+
+  /**
+   * Login attempt can have any form and defined by the specific provider.
+   */
+  value: unknown;
+}
+
+export interface AuthenticatorOptions {
+  config: Pick<ConfigType, 'sessionTimeout' | 'authc'>;
+  basePath: HttpServiceSetup['basePath'];
+  loggers: LoggerFactory;
+  clusterClient: PublicMethodsOf<ClusterClient>;
+  sessionStorageFactory: SessionStorageFactory<ProviderSession>;
+  isSystemAPIRequest: (request: KibanaRequest) => boolean;
+}
+
+// Mapping between provider key defined in the config and authentication
+// provider class that can handle specific authentication mechanism.
+const providerMap = new Map<
```

Comment:
> optional: ReadonlyMap
> question: if we use a dictionary, could it simplify typings, because all keys and values are known?

I plan to allow plugins that depend on security to extend provider list and hence add new providers to it, so we'll know the full mapping only at run-time